packages/filesystem/src/browser/file-service.ts

```diff
@@ -1108,6 +1108,37 @@ export class FileService {
         return fileStat;
     }
 
+    protected async incrementFileName(namePrefix: string, extSuffix: string): Promise<string> {
+        // name copy 5(.txt) => name copy 6(.txt)
+        // name copy(.txt) => name copy 2(.txt)
+        const suffixRegex = /^(.+ copy)( \d+)?$/;
+        if (suffixRegex.test(namePrefix)) {
+            return namePrefix.replace(suffixRegex, (match, g1?, g2?) => {
+                const number = (g2 ? parseInt(g2) : 1);
+                return number === 0
+                    ? `${g1}`
+                    : (number < 30
+                        ? `${g1} ${number + 1}`
+                        : `${g1}${g2} copy`);
+            }) + extSuffix as string;
+        }
+
+        // name(.txt) => name copy(.txt)
+        return `${namePrefix} copy${extSuffix}`;
+    }
+
+    private async findValidPasteFileTarget(target: URI): Promise<URI> {
+        let candidate: URI = target;
+        while (true) {
+            if (!await this.exists(candidate)) {
+                break;
+            }
+            const name = await this.incrementFileName(candidate.path.name, target.path.ext);
+            candidate = target.parent.resolve(name);
+        }
+        return candidate;
+    }
+
```

Comment:
@i501378 I believe the code will require a [CQ](https://github.com/eclipse-theia/theia/wiki/Registering-CQs) since it was copied/inspired by VS Code.

Reply:
yes. i will create it.

Reply:
no need, because the code is updated

Reply:
I removed the CQ Required label.